### PR TITLE
Feature/code for review

### PR DESCRIPTION
**Pull request by jason102**
Please see the code review spreadsheet so you can choose and know what parts to review [HERE](https://docs.google.com/spreadsheets/d/1baXEQGownokMN9KyGprjKGWsrroQEpsZaSl4bDVp60E/edit#gid=0)

(Although you're more than welcome to review all the PR code changes if you want!!) 

My recommendation is for you to clone the code to your computer and then check out this branch (feature/code-for-review) so you can review it offline and then leave comments here. But you're welcome to review it any way you prefer.

### CODE REVIEW

File: client/src/app/RefreshTokensContext.tsx [hunk 1]
```diff
@@ -0,0 +1,43 @@
+import React, { createContext, useEffect, useState, useRef } from "react";
+import { useAppDispatch, useAppSelector } from "src/app/store";
+import refreshToken from "src/app/api/auth/refreshTokenThunk";
+
+export const RefreshTokensContext = createContext<{
```
Comment (ndp) [inline client/src/app/RefreshTokensContext.tsx]: I'd be curious if Typescript can properly infer the type here  ... you could try leaving out the `{ hasRefreshToken: boolean }` and see if it still works... some example on the web imply it does.

File: client/src/app/RefreshTokensContext.tsx [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    const getRefreshedAccessToken = async () => {
+      await dispatch(refreshToken());
+      setHasRefreshedToken(true);
+    };
+
+    if (!accessToken && firstRender.current) {
+      firstRender.current = false;
+      getRefreshedAccessToken();
+    }
+  }, [accessToken, dispatch]);
```
Comment (ndp) [inline client/src/app/RefreshTokensContext.tsx:32]: I'm not that current on current React practices (I never liked hooks), but this seems weird to have all this code to defeat the double rendering Facebook put in to protect you. Since you just want it to run once, can you just leave out all the dependencies for this `useEffect` and get the same result?

File: client/src/app/api/apiErrorUtils.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
+import { FetchBaseQueryError } from "@reduxjs/toolkit/query";
+
+export const OOPS_MESSAGE =
+  "Oops, something went wrong! Please contact Jason to get it fixed!";
+
+// Copied from RTK docs for typing caught errors
+export function isFetchBaseQueryError(
+  error: unknown
+): error is FetchBaseQueryError {
+  return typeof error === "object" && error != null && "status" in error;
```
Comment (ndp) [inline client/src/app/api/apiErrorUtils.ts]: if `typeof error === "object"` it will not be null.

File: client/src/app/api/apiErrorUtils.ts [hunk 4]
```diff
@@ -0,0 +1,11 @@
+import { FetchBaseQueryError } from "@reduxjs/toolkit/query";
+
+export const OOPS_MESSAGE =
+  "Oops, something went wrong! Please contact Jason to get it fixed!";
+
+// Copied from RTK docs for typing caught errors
```
Comment (ndp) [inline client/src/app/api/apiErrorUtils.ts:6]: I like leaving the URL here, if possible. I'm not sure what RTK is.

File: client/src/pages/logged-in-pages/Location/ImageCarousel/index.tsx [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+        </Box>
+      )}
+      <LoadingDialog
+        isOpen={showDeleteConfirmDialog}
+        isLoading={isDeletingPhoto}
+        onConfirmButtonClick={onDeletePhoto}
+        title="Confirm Photo Deletion"
+        contentText="Are you sure you want to delete this photo?"
+        yesNoOption
+        onNoOptionClick={() => setShowDeleteConfirmDialog(false)}
```
Comment (ndp) [inline client/src/pages/logged-in-pages/Location/ImageCarousel/index.tsx:140]: This component is pretty long... is it possible to break off some of the pieces, like the upload control?  Just an idea.

File: client/src/pages/logged-in-pages/Location/ImageCarousel/photoUtils.ts [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+};
+
+export const convertImageFilesToBase64Strings = async (formData: FormData) => {
+  const file2Base64Promises = Array.from(formData.values()).map((file) =>
+    file2Base64(file as File)
+  );
+
+  const base64Files = await Promise.all(file2Base64Promises);
+
+  return base64Files;
```
Comment (ndp) [inline client/src/pages/logged-in-pages/Location/ImageCarousel/photoUtils.ts:91]: This looks good... seems like how you have to do it... 

The variable `base64Files` is redundant and could be removed... or renamed as `files` seems inaccurate.

File: client/src/app/api/apiErrorUtils.ts (hunk 3 above)
Comment (jason102) [reply client/src/app/api/apiErrorUtils.ts]: image
When I remove && error != null, TS complains about the last condition. I copied this from the official RTK docs, but I'll update it to use !== instead of !=

File: client/src/app/RefreshTokensContext.tsx (hunk 1 above)
Comment (jason102) [reply client/src/app/RefreshTokensContext.tsx]: Yes, you're correct! I removed the generic.

File: client/src/app/RefreshTokensContext.tsx (hunk 2 above)
Comment (jason102) [reply client/src/app/RefreshTokensContext.tsx:32]: I've done some digging online and it appears my approach of using a ref is currently what most people do. I can't find a github issues conversation I saw before recommending this, but here's what ChatGPT 4.0 says: 

In React, `StrictMode` is a tool for highlighting potential problems in an application. It intentionally double-invokes certain functions, including render methods and useEffect, to help you find side-effects in your render phase and prepare your code for the future concurrent mode. However, it's important to note that `StrictMode` only runs in development mode and has no impact in production.

When dealing with effects that you only want to run once (like network requests), it's common to see developers use a ref flag to ensure that the code inside `useEffect` doesn't run more than once, as you've done in your code. This approach is generally considered a practical workaround for the double invocation caused by `StrictMode`.

React's official documentation doesn't provide a specific recommendation for handling this `StrictMode` behavior because it's more a development aid than an aspect of React's core functionality that needs to be worked around. The ref-based approach you've used is a common pattern among React developers to prevent duplicate calls in development mode.

Regarding your question about returning a destructor function in the `useEffect` hook:

1. **Cancellation of Promises:** JavaScript promises, including the ones used in async/await syntax, are not cancellable. This means once a network request is fired off, there's no built-in way to stop it mid-flight. However, you can use mechanisms like AbortController with fetch API to cancel requests. If you're using another library like Axios, it may have its own cancellation mechanisms.

2. **Cleanup Functions in `useEffect`:** The cleanup function in `useEffect` is designed for cleaning up side-effects (like event listeners or clearing timeouts/intervals). If your network request needs cleanup (like unsubscribing from a subscription), you would return a cleanup function. But for a simple network request with no ongoing effects, it's not necessary.

3. **Best Practices:** If you're concerned about the effect running twice in development mode due to `StrictMode`, your current approach using `useRef` is a practical solution. Just remember that this is a development-mode issue, and in production, your effect will only run once as expected.

In summary, your approach with `useRef` is a commonly used and practical solution to the `StrictMode` double invocation in development mode. For cancellation and cleanup, use the cleanup function of `useEffect` only if your network request or its effects persist beyond the lifecycle of the component.

File: client/src/app/api/apiErrorUtils.ts (hunk 4 above)
Comment (jason102) [reply client/src/app/api/apiErrorUtils.ts:6]: Adding the URL! RTK is short for Redux Toolkit:
https://redux-toolkit.js.org/rtk-query/usage-with-typescript#inline-error-handling-example

File: client/src/pages/logged-in-pages/Location/ImageCarousel/index.tsx (hunk 5 above)
Comment (jason102) [reply client/src/pages/logged-in-pages/Location/ImageCarousel/index.tsx:140]: I agree this file can be shorter! I moved all the `onUploadPhoto()` code to its own hook in the same folder.

File: client/src/pages/logged-in-pages/Location/ImageCarousel/photoUtils.ts (hunk 6 above)
Comment (jason102) [reply client/src/pages/logged-in-pages/Location/ImageCarousel/photoUtils.ts:91]: Hmm yeah I know what you mean. Maybe it's just me but sometimes I prefer to do it this way and not directly return the awaited promise so it's a little more clear in the code exactly what it was that the promise resolved.
